[JSC] Implement `Promise.race` in C++

sosukesuzuki opened the pull request:
#### 08283a20d0ab1d86ef99406d6138df320033f4ea
<pre>
[JSC] Implement `Promise.race` in C++
<a href="https://bugs.webkit.org/show_bug.cgi?id=300569">https://bugs.webkit.org/show_bug.cgi?id=300569</a>

Reviewed by Yusuke Suzuki.

This patch changes to implement `Promise.race` in C++.

DFG/FTL enabled:
                            TipOfTree                  Patched

    promise-race          8.3883+-0.1417     ^      6.6330+-0.1397        ^ definitely 1.2646x faster

DFG/FTL disabled:
                            TipOfTree                  Patched

    promise-race         16.2104+-0.1943     ^      7.9515+-0.2155        ^ definitely 2.0387x faster

Tests: JSTests/microbenchmarks/promise-race.js
       JSTests/stress/promise-race-empty-args.js
       JSTests/stress/promsie-race-resolve-getter-throws.js

* JSTests/microbenchmarks/promise-race.js: Added.
* JSTests/stress/new-promise-capabilities-requires-constructor.js:
(shouldThrow):
* JSTests/stress/promise-race-empty-args.js: Added.
(shouldBe):
* JSTests/stress/promsie-race-resolve-getter-throws.js: Added.
(shouldBe):
* Source/JavaScriptCore/builtins/PromiseConstructor.js:
(any):
(race): Deleted.
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::promiseResolveWatchpointSet):
* Source/JavaScriptCore/runtime/JSPromiseConstructor.cpp:
(JSC::isFastPromiseConstructor):
(JSC::promiseRaceSlow):
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: <a href="https://commits.webkit.org/302567@main">https://commits.webkit.org/302567@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4f43f5fefcf7ed687b2aa294cb24113c54f9351f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/129508 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1765 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/40347 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/136891 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/80936 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/1697 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/1641 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/98648 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/80936 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/132455 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/1316 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/115999 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/79295 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/1237 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/34130 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/80165 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/121500 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/109707 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34628 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/139365 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/127960 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/1555 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/1491 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/107171 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/1597 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/112340 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/107012 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27249 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/1267 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/30858 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/54247 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1626 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/64989 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/160974 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/1446 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/40153 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/1480 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/1548 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/runtime/JSPromiseConstructor.cpp:
```diff
@@ -102,4 +106,96 @@ JSC_DEFINE_HOST_FUNCTION(promiseConstructorFuncWithResolvers, (JSGlobalObject* g
+    auto [promise, resolve, reject] = JSPromise::newPromiseCapability(globalObject, thisValue);
+    RETURN_IF_EXCEPTION(scope, { });
+
+    auto callReject = [&]() -> void {
```
We should probably have a super fast path when |this| is Promise. So we can just use JSPromise* and JSPromise::reject / JSPromise::resolve. See Promise.all JS code as a reference.

Constellation commented on Source/JavaScriptCore/runtime/JSPromiseConstructor.cpp:
```diff
@@ -102,4 +106,96 @@ JSC_DEFINE_HOST_FUNCTION(promiseConstructorFuncWithResolvers, (JSGlobalObject* g
+
+        JSValue then = nextPromise.get(globalObject, vm.propertyNames->then);
+        RETURN_IF_EXCEPTION(scope, void());
+        if (auto* nextPromiseObj = jsDynamicCast<JSPromise*>(nextPromise); nextPromiseObj && nextPromiseObj->isThenFastAndNonObservable()) [[likely]] {
```
You can skip then property lookup if the condition is met

Constellation commented on Source/JavaScriptCore/runtime/JSPromiseConstructor.cpp:
```diff
@@ -102,4 +106,93 @@ JSC_DEFINE_HOST_FUNCTION(promiseConstructorFuncWithResolvers, (JSGlobalObject* g
+    if (!thisValue.isObject()) [[unlikely]]
+        return throwVMTypeError(globalObject, scope, "|this| is not an object"_s);
+
+    auto [promise, resolve, reject] = JSPromise::newPromiseCapability(globalObject, thisValue);
```
We should avoid doing this when |this| is Promise since this is allocating unnecessary functions

sosukesuzuki replied:
Even if `this` were a `Promise`, wouldn't it be necessary to call `promise->performThen` within `forEachInIteration`, and pass `resolve` and `reject` as arguments, thus requiring allocation of these functions?

```cpp
// fast path code example
    forEachInIterable(globalObject, iterable, [&](VM& vm, JSGlobalObject* globalObject, JSValue value) {
        auto scope = DECLARE_THROW_SCOPE(vm);

        JSPromise* nextPromise = JSPromise::resolvedPromise(globalObject, value);
        RETURN_IF_EXCEPTION(scope, void()); 

        scope.release();
        nextPromise->performPromiseThen(globalObject, resolve, reject, jsUndefined(), promise);
    });
```

Constellation replied:
No. In that case, you can pass both undefined for handlers. Then since both are promises, these functions are invisible so we can just use the shared functions.

Constellation replied:
But in that case we also need a bit careful handling of then function (see Promise.all code, it is allocating a special callbacks only when then is not a default then (since it becomes observable).

But this case requires a bit careful checks (then is default one). However, since it will be 99.9999% cases, definitely we should have that optimization. Maybe it would be OK for the initial version, but FIXME is necessary and we need to fix it.

Constellation commented on Source/JavaScriptCore/runtime/JSPromiseConstructor.cpp:
```diff
@@ -102,4 +106,93 @@ JSC_DEFINE_HOST_FUNCTION(promiseConstructorFuncWithResolvers, (JSGlobalObject* g
+        call(globalObject, reject, rejectCallData, jsUndefined(), rejectArguments);
+    };
+
+    JSValue promiseResolveValue = thisValue.get(globalObject, vm.propertyNames->resolve);
```
Ditto. We should do promiseResolve ops in C++ efficiently when |this| is Promise

sosukesuzuki commented on Source/JavaScriptCore/runtime/JSPromiseConstructor.cpp:
```diff
@@ -102,4 +108,156 @@ JSC_DEFINE_HOST_FUNCTION(promiseConstructorFuncWithResolvers, (JSGlobalObject* g
+
+        if (nextPromise->isThenFastAndNonObservable()) [[likely]] {
+            scope.release();
+            nextPromise->performPromiseThen(globalObject, resolve, reject, jsUndefined(), promise);
```
@Constellation  If pass `jsUndefined()` instead of resolve and reject here, the following test262 will fail.
```
! NEW FAIL test/built-ins/Promise/race/S25.4.4.3_A6.2_T1.js (strict mode) (Exit code: 3)
Test262:AsyncTestFailure:Test262Error: Test262Error: The value of sequence.length is expected to be 5 Expected SameValue(«4», «5») to be true

! NEW FAIL test/built-ins/Promise/race/S25.4.4.3_A6.2_T1.js (default) (Exit code: 3)
Test262:AsyncTestFailure:Test262Error: Test262Error: The value of sequence.length is expected to be 5 Expected SameValue(«4», «5») to be true

! NEW FAIL test/built-ins/Promise/race/S25.4.4.3_A6.1_T1.js (strict mode) (Exit code: 3)
Test262:AsyncTestFailure:Test262Error: Test262Error: Expected SameValue(«4», «5») to be true

! NEW FAIL test/built-ins/Promise/race/S25.4.4.3_A6.1_T1.js (default) (Exit code: 3)
Test262:AsyncTestFailure:Test262Error: Test262Error: Expected SameValue(«4», «5») to be true
```

Constellation replied:
Did you also pass the resolution promise instead of `jsUndefined()` for the 4th parameter?

Constellation replied:
BTW, this means that this `jsUndefined()` handler-less thing only works when `promise` is actually `JSPromise*`. And that's why I mentioned that we can do this optimization only when `|this|` is `Promise`. (And `Promise.all` is checking it at the prologue for the fast path).

sosukesuzuki replied:
When the 4th parameter is set to `promise`, the following debug assertion error occurs. if assertions are disabled, the test passes.
```
! NEW FAIL test/built-ins/Promise/race/resolved-sequence-mixed.js (default) (Exit code: 133)
ASSERTION FAILED: status() == Status::Pending
./runtime/JSPromise.cpp(327) : void JSC::JSPromise::fulfillPromise(JSGlobalObject *, JSValue)
1   0x10e963cdc JSC::JSPromise::fulfillPromise(JSC::JSGlobalObject*, JSC::JSValue)
2   0x10e963598 JSC::JSPromise::resolvePromise(JSC::JSGlobalObject*, JSC::JSValue)
3   0x10e925c5c JSC::runInternalMicrotask(JSC::JSGlobalObject*, JSC::InternalMicrotask, std::__1::span<JSC::JSValue const, 4ul>)
4   0x10ed91d38 JSC::VM::drainMicrotasks()::$_0::operator()(JSC::QueuedTask&) const
5   0x10ed653d4 void JSC::MicrotaskQueue::performMicrotaskCheckpoint<JSC::VM::drainMicrotasks()::$_0>(JSC::VM&, JSC::VM::drainMicrotasks()::$_0 const&)
6   0x10ed65160 JSC::VM::drainMicrotasks()
7   0x104135bd8 int runJSC<jscmain(int, char**)::$_1>(CommandLine const&, bool, jscmain(int, char**)::$_1 const&)
8   0x1041325b0 jscmain(int, char**)
9   0x104132098 main
10  0x189ff6b98 start

sh: line 1: 97906 Trace/BPT trap: 5       DYLD_FRAMEWORK_PATH=/Users/sosukesuzuki/ghq/github.com/WebKit/WebKit/WebKitBuild/Debug TZ=PST /Users/sosukesuzuki/ghq/github.com/WebKit/WebKit/WebKitBuild/Debug/jsc --test262-async /var/folders/hh/jdy4bb8d6dj0xn02gwvt58zh0000gn/T/MxnXsOReEJ/fEXmN3JqBi /var/folders/hh/jdy4bb8d6dj0xn02gwvt58zh0000gn/T/MxnXsOReEJ/XxzyxXJMaB --strict-file=/Users/sosukesuzuki/ghq/github.com/WebKit/WebKit/JSTests/test262/test/built-ins/Promise/race/S25.4.4.3_A7.3_T1.js 2>&1
! NEW FAIL test/built-ins/Promise/race/S25.4.4.3_A7.3_T1.js (strict mode) (Exit code: 133)
ASSERTION FAILED: status() == Status::Pending
./runtime/JSPromise.cpp(304) : void JSC::JSPromise::rejectPromise(JSGlobalObject *, JSValue)
1   0x10d1cb8f0 JSC::JSPromise::rejectPromise(JSC::JSGlobalObject*, JSC::JSValue)
2   0x10d18d9d8 JSC::runInternalMicrotask(JSC::JSGlobalObject*, JSC::InternalMicrotask, std::__1::span<JSC::JSValue const, 4ul>)
3   0x10d5f9d38 JSC::VM::drainMicrotasks()::$_0::operator()(JSC::QueuedTask&) const
4   0x10d5cd3d4 void JSC::MicrotaskQueue::performMicrotaskCheckpoint<JSC::VM::drainMicrotasks()::$_0>(JSC::VM&, JSC::VM::drainMicrotasks()::$_0 const&)
5   0x10d5cd160 JSC::VM::drainMicrotasks()
6   0x102a95bd8 int runJSC<jscmain(int, char**)::$_1>(CommandLine const&, bool, jscmain(int, char**)::$_1 const&)
7   0x102a925b0 jscmain(int, char**)
8   0x102a92098 main
9   0x189ff6b98 start
```

sosukesuzuki replied:
I read the code and removed the assertion status() == Status::Pending, as I thought it was being called intentionally on a promise that is not pending. Is this safe?

Constellation commented on Source/JavaScriptCore/runtime/JSPromiseConstructor.cpp:
```diff
@@ -125,4 +131,163 @@ JSC_DEFINE_HOST_FUNCTION(promiseConstructorFuncWithResolvers, (JSGlobalObject* g
+            if (!resolve || !reject) {
+                auto [res, rej] = promise->createFirstResolvingFunctions(vm, globalObject);
+                resolve = res;
+                reject = rej;
```
You should use
```
std::tie(resolve, reject) = promise->createFirstResolvingFunctions(vm, globalObject);
```

Constellation commented on Source/JavaScriptCore/runtime/JSPromiseConstructor.cpp:
```diff
@@ -125,4 +131,163 @@ JSC_DEFINE_HOST_FUNCTION(promiseConstructorFuncWithResolvers, (JSGlobalObject* g
+    if (!isFastPromiseConstructor(globalObject, thisValue)) [[unlikely]] {
+        scope.release();
+        return JSValue::encode(promiseRaceSlow(globalObject, callFrame, thisValue));
+    }
```
Do,
`RELEASE_AND_RETURN(scope, JSValue::encode(promiseRaceSlow(globalObject, callFrame, thisValue)));`

Constellation commented on Source/JavaScriptCore/runtime/JSPromiseConstructor.cpp:
```diff
@@ -125,4 +131,163 @@ JSC_DEFINE_HOST_FUNCTION(promiseConstructorFuncWithResolvers, (JSGlobalObject* g
+    }
+
+    auto* promise = JSPromise::create(vm, globalObject->promiseStructure());
+    RETURN_IF_EXCEPTION(scope, { });
```
This is not necessary as it never fails.

Constellation commented on Source/JavaScriptCore/runtime/JSPromiseConstructor.cpp:
```diff
@@ -125,4 +131,163 @@ JSC_DEFINE_HOST_FUNCTION(promiseConstructorFuncWithResolvers, (JSGlobalObject* g
+
+    if (!promiseResolveValue.isCallable()) [[unlikely]] {
+        throwTypeError(globalObject, scope, "Promise resolve is not a function"_s);
+        callReject();
```
This looks wrong as we cannot execute a code after error is thrown.
We should just reject the promise with `createTypeError` value.

Constellation commented on Source/JavaScriptCore/runtime/JSPromiseConstructor.cpp:
```diff
@@ -125,4 +131,159 @@ JSC_DEFINE_HOST_FUNCTION(promiseConstructorFuncWithResolvers, (JSGlobalObject* g
+                std::tie(resolve, reject) = promise->createFirstResolvingFunctions(vm, globalObject);
+            JSValue then = nextPromise->get(globalObject, vm.propertyNames->then);
+            RETURN_IF_EXCEPTION(scope, void());
+            CallData thenCallData = getCallData(then);
```
Lets use `getCallDataInline`.

Constellation commented on Source/JavaScriptCore/runtime/JSPromiseConstructor.cpp:
```diff
@@ -125,4 +131,159 @@ JSC_DEFINE_HOST_FUNCTION(promiseConstructorFuncWithResolvers, (JSGlobalObject* g
+    VM& vm = globalObject->vm();
+    auto* object = asObject(value);
+    if (object->getDirectOffset(vm, vm.propertyNames->resolve) != invalidOffset) [[unlikely]]
+        return false;
```
I think we should have a watchpoint for detecting that `resolve` is not changed.

Constellation commented on Source/JavaScriptCore/runtime/JSPromise.cpp:
```diff
@@ -287,7 +287,6 @@ void JSPromise::performPromiseThen(VM& vm, JSGlobalObject* globalObject, JSValue
 
 void JSPromise::rejectPromise(VM& vm, JSGlobalObject* globalObject, JSValue argument)
 {
-    ASSERT(status() == Status::Pending);
```
This is actually not correct since this function should be used only after the first resolving call gets done already.

Constellation commented on Source/JavaScriptCore/runtime/JSPromise.cpp:
```diff
@@ -303,7 +302,6 @@ void JSPromise::rejectPromise(VM& vm, JSGlobalObject* globalObject, JSValue argu
 
 void JSPromise::fulfillPromise(VM& vm, JSGlobalObject* globalObject, JSValue argument)
 {
-    ASSERT(status() == Status::Pending);
```
Ditto. You cannot override the promise again.

Constellation commented on Source/JavaScriptCore/runtime/JSPromiseConstructor.cpp:
```diff
@@ -125,4 +131,159 @@ JSC_DEFINE_HOST_FUNCTION(promiseConstructorFuncWithResolvers, (JSGlobalObject* g
+
+        if (nextPromise->isThenFastAndNonObservable()) [[likely]] {
+            scope.release();
+            nextPromise->performPromiseThen(vm, globalObject, jsUndefined(), jsUndefined(), promise, promise);
```
I think this needs a bit more work here. You probably need to add different InternalMicrotask, adding `performPromiseThen` variant which takes microtask id instead of handlers etc., but if it is too complicated right now, then it is fine for now to generate `resolve` / `reject` function and passing them here (while it is a bit costly still).

Constellation commented on Source/JavaScriptCore/runtime/JSPromiseConstructor.cpp:
```diff
@@ -52,7 +57,7 @@ const ClassInfo JSPromiseConstructor::s_info = { "Function"_s, &Base::s_info, &p
   resolve         promiseConstructorFuncResolve        DontEnum|Function 1 PromiseConstructorResolveIntrinsic
   reject          promiseConstructorFuncReject         DontEnum|Function 1 PromiseConstructorRejectIntrinsic
-  race            JSBuiltin                            DontEnum|Function 1
+  race            promiseConstructorFuncRace                            DontEnum|Function 1
```
Let's remove spaces before `DontEnum|Function 1` to align to the other definitions.

Constellation reviewed (approved):
r=me

sosukesuzuki committed:
[JSC] Implement `Promise.race` in C++
https://bugs.webkit.org/show_bug.cgi?id=300569

Reviewed by Yusuke Suzuki.

This patch changes to implement `Promise.race` in C++.

DFG/FTL enabled:
                            TipOfTree                  Patched

    promise-race          8.3883+-0.1417     ^      6.6330+-0.1397        ^ definitely 1.2646x faster

DFG/FTL disabled:
                            TipOfTree                  Patched

    promise-race         16.2104+-0.1943     ^      7.9515+-0.2155        ^ definitely 2.0387x faster

Tests: JSTests/microbenchmarks/promise-race.js
       JSTests/stress/promise-race-empty-args.js
       JSTests/stress/promsie-race-resolve-getter-throws.js

* JSTests/microbenchmarks/promise-race.js: Added.
* JSTests/stress/new-promise-capabilities-requires-constructor.js:
(shouldThrow):
* JSTests/stress/promise-race-empty-args.js: Added.
(shouldBe):
* JSTests/stress/promsie-race-resolve-getter-throws.js: Added.
(shouldBe):
* Source/JavaScriptCore/builtins/PromiseConstructor.js:
(any):
(race): Deleted.
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::promiseResolveWatchpointSet):
* Source/JavaScriptCore/runtime/JSPromiseConstructor.cpp:
(JSC::isFastPromiseConstructor):
(JSC::promiseRaceSlow):
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: https://commits.webkit.org/302567@main
